packages/teleport/src/Discover/DownloadScript/useDownloadScript.tsx

```diff
@@ -22,20 +22,120 @@ import { AgentStepProps } from '../types';
+
+    // Set flags to default.
+    setRetry(false);
+    setPollState('polling');
+
+    // abortController is required to cancel any in flight request.
+    const abortController = new AbortController();
+    const abortSignal = abortController.signal;
+    let timeoutId;
+    let intervalId;
+
+    // inFlightReq is a flag to prevent another fetch request when a
+    // previous fetch request is still in progress. May happen when a
+    // previous fetch request is taking longer than the polling
+    // interval time.
+    let inFlightReq;
+
+    function fetchNodeMatchingRefResourceId(token: JoinToken) {
+      if (inFlightReq) return;
+
+      inFlightReq = ctx.nodesService
+        .fetchNodes(
+          ctx.clusterId,
+          {
+            search: `${token.refResourceId}`,
+            limit: 1,
+          },
+          abortSignal
+        )
+        .then(res => {
+          if (res.agents.length > 0) {
+            setPollState('success');
+            props.updateAgentMeta({
+              ...props.agentMeta,
+              refResourceId: token.refResourceId,
+              node: res.agents[0],
+            });
+            cleanUp();
+          }
+        })
+        // Polling related errors are ignored.
+        // The most likely cause of error would be network issues
+        // and aborting in flight request.
+        .catch(() => {})
+        .finally(() => {
+          inFlightReq = null; // reset flag
+        });
+    }
+
+    function cleanUp() {
+      clearInterval(intervalId);
+      clearTimeout(timeoutId);
+      // Cancel any in flight request.
+      abortController.abort();
+    }
+
     run(() =>
-      ctx.joinTokenService.fetchJoinToken(['Node'], 'token').then(setJoinToken)
+      ctx.joinTokenService.fetchJoinToken(['Node'], 'token').then(token => {
+        // Probably will never happen, but just in case, otherwise
+        // querying for the resource can return a false positive.
+        if (!token.refResourceId) {
+          setAttempt({
+            status: 'failed',
+            statusText:
+              'reference resource ID is required to discover the newly added resource',
+          });
+          return;
+        }
+
+        setJoinToken(token);
+
+        // Start the poller to discover the resource just added.
+        intervalId = setInterval(
+          () => fetchNodeMatchingRefResourceId(token),
+          THREE_SECONDS_IN_MS
+        );
+
+        // Set a timeout in case polling continuosly produces
+        // no results. Which means there is either a network error,
+        // script is ran unsuccessfully, script has not been ran,
+        // or resource cannot connect to cluster.
+        timeoutId = setTimeout(() => {
+          setPollState('error');
+          cleanUp();
+        }, FIVE_MINUTES_IN_MS);
+      })
     );
-  }, []);
+
+    return () => {
+      cleanUp();
+    };
+  }, [retry]);
+
+  function regenerateScriptAndRepoll() {
```

Comment:
Using a 'flag' variable for a hook is an anti-pattern. Hook dependencies should only be values if you can help it.

In this case what I think you can do instead is to have this function call to generate the new token, then set that token into state and use that token as the dependent for the polling call. 

```ts
  function regenerateScriptAndRepoll() {
    const joinToken = fetchJoinToken();
    setJoinToken(joinToken)
  }
```